### ISSUE TITLE
[Visual bug] Caustics don't show when other surface is with low roughness

### ISSUE THREAD

**Issue by ladna373** [TU-Clausthal-Rendering/ReSTIR-FG#1]
based on the name itself, the problem is that there is no caustic on another surface like this

**Comment by ladna373** [comment]
![low](https://github.com/user-attachments/assets/324917c8-cd8b-4401-98c2-dc68991ec706)
![high](https://github.com/user-attachments/assets/af9aa84b-b467-4a24-bd9e-b16eacef678c)

**Comment by SirKero** [comment]
That is expected behavior, as caustics cannot be displayed on surfaces where the roughness is too low, as the BRDF lobes get too thin and the BRDF evaluates to 0.
Because we use a fixed surface definition, a surface is either classified as diffuse or specular, Photons are only stored and collected on diffuse surfaces. The photons use the same definition, as we used to determine the first hit point. The values can be changed in this setting:
![screen1](https://github.com/user-attachments/assets/b70ebcf7-3a4e-428e-9861-bd5c25df5015)

However, it should be noted, that a roughness value lower than 0.08 uses delta functions (perfect mirror), so BRDF for caustics is always 0 at that point. Falcors BSDF can also be a bit weird at times, especially with low-roughness materials that have no metallic parts.
Also, this scene uses spotlights, with emissive lights the path tracer would hit the light source at the caustic spots, as now the reflection is pretty well defined.

Still, if the cutoff is set down, caustics are visible as long as the wall has a roughness of >=0.08. However, correlation artifacts will appear for resampling if the cutoff is set that low. Here is an example where the roughness of the wall is set to 0.08 with the cutoff settings from above:
![rough0_08](https://github.com/user-attachments/assets/30593e6e-c6dc-43c9-b799-ba01477f402c)
(Note that, indirect light is now pretty noisy for these surfaces)

To handle such low-roughness surfaces better, we would probably need to implement Hybrid Shift Mapping from the GRIS paper to acquire the first diffuse surface

**Comment by ladna373** [comment]
@SirKero thank you very much for the answer and for your help! The result may be “noisy”, but even so, it’s very impressive! I wanted to ask how long it will take to implement HSM in RESTIR FG?

**Comment by SirKero** [comment]
Thanks!

We currently have no ETA for HSM as I am currently busy with other research. But it would be highly interesting how well it actually works and how much it affects performance, as HSM involves tracing more rays.

**Comment by ladna373** [comment]
maybe it's too late to write something, but I wanted to ask - if the renderer is real time, then how to implement hsm? i heard its working only on "interactive" renderers

**Comment by SirKero** [comment]
HSM is not necessarily "interactive", but is definitely slower than tracing a random ray.
In our case, we would only use temporal HSM to the first diffuse hit (similar to SuffixReSTIR). While this will be slower, as now two paths need to be traced per pixel, the increase should not be so drastic that real-time is no longer possible.

**Comment by ladna373** [comment]
Thank you very much for your answer! And how will temporal differ from regular?

**Comment by SirKero** [comment]
You can use HSM to resample spatiotemporally (as used in ReSTIR PT). However, this is for one more costly (one or more spatial samples that need to be retraced) and is very prone to correlations.  One additional temporal sample to trace to the first diffuse surface should be sufficient, as spatiotemporal ReSTIR FG would be used on the first diffuse hit.

**Comment by ladna373** [comment]
Thanks again for the answer, I really appreciate it! I would also like to ask: will there be an update soon with this feature added?

**Comment by SirKero** [comment]
Probably not anytime soon. 
Adding HSM is not that simple and it is currently not that high on our priority list.

**Comment by ladna373** [comment]
can I find out what is your priority for adding to the restir or will you not update the restir fg at all?

**Comment by SirKero** [comment]
ReSTIR FG is mostly in maintenance mode, so don't expect any huge feature updates.
Big features like HSM will only be added if circumstances allow it (e.g. needed or of interest for future research) and I cannot promise if they will ever happen.

**Comment by ladna373** [comment]
thank you for the answer! Sorry for another question, but will indirect illumination emit caustics?

**Comment by SirKero** [comment]
Yes, indirect illumination can create caustics. This can also be observed in the Veach Ajar scene.
There is also a setting that can set the number of diffuse bounces a photon is allowed to make to still be classified as a caustic photon (set to 10 by default). 
![CausticBounces](https://github.com/user-attachments/assets/125164c8-46bb-4616-a650-e1396806f7f2)

**Comment by ladna373** [comment]
Thanks for the answer, it makes me incredibly happy! I also wanted to ask, will features that can change the behavior of caustics be added over time?

**Comment by SirKero** [comment]
I am unsure what you mean by "change the behavior" of caustics.
Progressive radius reduction (SPPM) and an alternative temporal collection mode are currently implemented for caustics and currently, there is nothing else planned for caustics.

**Comment by ladna373** [comment]
and one more question about hsm: does it affect only correlations or also whether caustics will be displayed on a glossy surface?

**Comment by SirKero** [comment]
It will probably not affect caustics on glossy surfaces directly, however, reflected caustics should improve in quality, as hsm should guide the diffuse hit where the photons are collected to similar surfaces for neighboring pixels.

**Comment by ladna373** [comment]
so, if the roughness is below 0.08, then it will not display caustics even with hsm, right?

**Comment by SirKero** [comment]
yes

**Comment by SirKero** [comment]
This problem is now addressed by our follow up work, which can be found here:
https://github.com/TU-Clausthal-Rendering/Guided-ReSTIR-FG-Plus
